type_test.go

```diff
@@ -0,0 +1,564 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package httprequest
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+
+	jc "github.com/juju/testing/checkers"
+	"github.com/julienschmidt/httprouter"
+	gc "gopkg.in/check.v1"
+)
+
+type testSuite struct{}
+
+var _ = gc.Suite(&testSuite{})
+
+func body(s string) io.ReadCloser {
+	return ioutil.NopCloser(strings.NewReader(s))
+}
+
+var unmarshalTests = []struct {
+	about       string
+	val         interface{}
+	expect      interface{}
+	params      Params
+	expectError string
+}{{
+	about: "struct with simple fields",
+	val: struct {
+		F1          int    `httprequest:",form"`
+		F2          int    `httprequest:",form"`
+		G1          string `httprequest:",url"`
+		G2          string `httprequest:",url"`
+		H           string `httprequest:",body"`
+		UnknownForm string `httprequest:",form"`
+		UnknownPath string `httprequest:",url"`
+	}{
+		F1: 99,
+		F2: -35,
+		G1: "g1 val",
+		G2: "g2 val",
+		H:  "h val",
+	},
+	params: Params{
+		Request: &http.Request{
+			Form: url.Values{
+				"F1": {"99"},
+				"F2": {"-35", "not a number"},
+			},
+			Body: body(`"h val"`),
+		},
+		Path: httprouter.Params{{
+			Key:   "G2",
+			Value: "g2 val",
+		}, {
+			Key:   "G1",
+			Value: "g1 val",
+		}, {
+			Key:   "G1",
+			Value: "g1 wrong val",
+		}},
+	},
+}, {
+	about: "struct with renamed fields",
```

Comment:
you can't rename a body field - the tag name is just ignored. My current feeling is that's OK (and it's a hassle to check :))